File: Eclipse Projects/webapps/TeleDirectPhp/WebContent/scripts/php/DatabaseProvider.php
<?php
require_once 'config.php';
require_once 'PersonCollection.php';
require_once 'FieldMap.php';
//include ('FieldMap.php');

$file =  basename(__FILE__ ).': ';

class DatabaseProvider{
	private $database;
	
	function __construct(){
		global $file;
		$this->database = new PDO('mysql:host='.DB_SERVER.';dbname='.DB_DATABASE, DB_USERNAME, DB_PASSWORD);
		// set the error mode to exceptions
		$this->database->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
		error_log($file."Connected to database\n", 3, PHP_SCRIPT_PATH."debug.log");
	}
	
	function getDatabase(){
		return $this->database;
	}
	
	/**
	 * Fetches the list of persons and their details from the database.
	 * The results are added to PersonCollection object and sent into a response to the client
	 */
	function fetchPersonList(){
		global $file;
	
		$pCollection = new PersonCollection();
		try
		{
	
			$sql = $this->generateFetchQuery();
			$stmt = @$this->database->prepare($sql);
	
			// execute the prepared statement
			@$stmt->execute();
	
			$numResults = $stmt->rowCount();
			$rows = array();
	
			// if no contacts are found
			if($numResults == 0){
				$status = 'failure';
				$message = 'No contacts found';
			}elseif ($numResults > 0){
				$status = 'success';
				$message = $numResults .' contacts found';
			}
	
			/* Add each row to rows one at a time */
			while($row = $stmt->fetch(PDO::FETCH_ASSOC)){
				$rows[] = $row;
				error_log($file.json_encode($row), 3, PHP_SCRIPT_PATH."debug.log");
			}
				
			// add to collection
			$pCollection->add($rows);
		}
		catch(Exception $e)
		{
			$status = 'error';
			// if we are here, something has gone wrong with the database
			$message = $e->getMessage() .' We are unable to process your request. Please try again later.';
		}
	
		return json_encode(array(
				'status' => $status,
				'message' => $message,
				'data' => array(
						'persons' => $pCollection->get()
					)
			));
	}
	
	
	/**
	 * Generates a fetch query based on the attributes defined in the $personDB of FieldMap
	 */
	
	function generateFetchQuery(){
	
		global $PersonDB, $PersonJS;
	
		// generate select fields string
		$selectFields = '';
		foreach ($PersonDB as $attr => $string){
			if($attr == 'ID')
				$selectFields .= 'persons.'.$PersonDB->$attr.' AS '.$PersonJS->$attr.', ';
			else
				$selectFields .= $PersonDB->$attr.' AS '.$PersonJS->$attr.', ';
		}
		$selectFields = substr($selectFields, 0, -2);
	
		//create the query
		$sql = 'SELECT '.$selectFields.' FROM persons INNER JOIN personadddetails '.
				'ON persons.'.$PersonDB->ID.' = personadddetails.'.$PersonDB->ID.//;
				' ORDER BY '.$PersonJS->FIRST_NAME;
		
		error_log($file."fecth query = ".$sql." \n", 3, "debug.log");
		
		return $sql;
	}
	
	
	/**
	 * Used to save the changes made to the persons on the client side.
	 * The SQL procedure is called to add the details of the person.
	 * The array of persons is iterated through and the person details are extracted into $values
	 * which is then passed to the SQL procedure to add the corresponding person to the database 
	 */
	function saveToDatabase($persons){
		global $file, $PersonDB, $PersonJS;
		$status ='';
		$message= '';
	
		if($persons){
			try{	
	
				$sql = "CALL INITIATE_ADD()";
				$stmt = $this->database->prepare($sql);
				$stmt->execute();
				
				//prepare query
				$sql = "CALL ADD_PERSON(?,?,?,?,?,?,?,?,?,?,?,?,?,?)";
				$stmt = $this->database->prepare($sql);
					
				foreach($persons as $person){
					error_log($file."Got person: ".json_encode($person)."\n", 3, "debug.log");
	
					$values = array();
					foreach($person as $attr => $value){
						if($attr != $PersonJS->ID)
							array_push($values, $value);
					}
	
					error_log($file."Adding Person : ".implode(" ", $values).PHP_EOL, 3, "debug.log");
	
					/*** execute the prepared statement ***/
					$success = $stmt->execute($values);
	
					if($success){
						/** added succesfully **/
						$status = 'success';
					}else{
						/** there is some problem **/
						//throw new Exception("Could not add : ".implode(" ", $values));
						$message .= "Could not add : ".implode(" ", $values).PHP_EOL;
					}
	
				}
				$message .= "Saving Completed".PHP_EOL;
	
			}
			catch(Exception $e)
			{
				$status = 'error';
					
				/*** if we are here, something has gone wrong with the database ***/
				$message = $message .'We are unable to process your request. Please try again later.';
				$message = $message.' '.($e->getMessage());
				error_log($file."Error message ".$e->getMessage()." \n", 3, "debug.log");
			}
		}
	
		return json_encode(array(
				'status' => $status,
				'message' => $message
		));
	}
	
	
	/**
	 * Used to add the persons to the database.
	 * The SQL procedure is called to add the details of the person.
	 * The array of persons is iterated through and the person details are extracted into $values
	 * which is then passed to the SQL procedure to add the corresponding person to the database
	 */
	function addPersons($persons){
		global $file, $PersonDB, $PersonJS;
		$status ='';
		$message= '';
	
		if($persons){
			try{
				//prepare query
				$sql = "CALL ADD_PERSON(?,?,?,?,?,?,?,?,?,?,?,?,?,?)";
				$stmt = $this->database->prepare($sql);
					
				foreach($persons as $person){
					error_log($file."Got person: ".json_encode($person)."\n", 3, "debug.log");
	
					$values = array();
					foreach($person as $attr => $value){
						if($attr != $PersonJS->ID)
							array_push($values, $value);
					}
	
					error_log($file."Adding Person : ".implode(" ", $values).PHP_EOL, 3, "debug.log");
	
					/*** execute the prepared statement ***/
					$success = $stmt->execute($values);
	
					if($success){
						/** added succesfully **/
						$status = 'success';
					}else{
						/** there is some problem **/
						//throw new Exception("Could not add : ".implode(" ", $values));
						$message .= "Could not add : ".implode(" ", $values).PHP_EOL;
					}
	
				}
				$message .= "Adding Completed".PHP_EOL;
	
			}
			catch(Exception $e)
			{
				$status = 'error';
					
				/*** if we are here, something has gone wrong with the database ***/
				$message = $message .'We are unable to process your request. Please try again later.';
				$message = $message.' '.($e->getMessage());
				error_log($file."Error message ".$e->getMessage()." \n", 3, "debug.log");
			}
		}
	
		return json_encode(array(
				'status' => $status,
				'message' => $message
		));
	}
	/**
	 * Used to delete the persons from the database
	 * identified by the 'ids' given in the $ids object
	 */
	function deletePersons($ids){
		global $file;
		$status ='';
		$message= '';
	
		if($ids){
			try{
				$success = true;
	
				foreach($ids as $id){
					
					$sql = "CALL DELETE_PERSON(:id)";
					$stmt = $this->database->prepare($sql);
					$success &= $stmt->execute(array(':id' => $id));
					
					if($success){
					
						/** deleted succesfully **/
						$status = 'success';
					
					}else{
						/** there is some problem **/
						//throw new Exception("Could not add : ".implode(" ", $values));
						$message .= "Could not delete person id: ".$id.PHP_EOL;
					}
	
				}
				$message .= "Deletion Completed".PHP_EOL;
	
			}
			catch(Exception $e)
			{
				$status = 'error';
					
				/*** if we are here, something has gone wrong with the database ***/
				$message = $message .'We are unable to process your request. Please try again later.';
				$message = $message.' '.($e->getMessage());
				error_log($file."Error message ".$e->getMessage()." \n", 3, "debug.log");
			}
		}
	
		return json_encode(array(
				'status' => $status,
				'message' => $message
		));
	}
	
	/**
	 * Used to update the details of the persons
	 * sent in the persons object
	 */
	function updatePersons($persons){
		global $file, $PersonDB, $PersonJS;
		$status ='';
		$message= '';
	
		if($persons){
			try{	
	
				
				$sql = "CALL UPDATE_PERSON(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)";
				$stmt = $this->database->prepare($sql);
					
				foreach($persons as $person){
					error_log($file."Got person: ".json_encode($person)."\n", 3, "debug.log");
	
					$values = array();
					foreach($person as $attr => $value){
						array_push($values, $value);
					}
	
					error_log($file."Updating Person : ".implode(" ", $values).PHP_EOL, 3, "debug.log");
	
					/*** execute the prepared statement ***/
					$success = $stmt->execute($values);
	
					if($success){
						/** added succesfully **/
						$status = 'success';
					}else{
						/** there is some problem **/
						//throw new Exception("Could not add : ".implode(" ", $values));
						$message .= "Could not update : ".implode(" ", $values).PHP_EOL;
					}
	
				}
				$message .= "Updating Completed".PHP_EOL;
	
			}
			catch(Exception $e)
			{
				$status = 'error';
					
				/*** if we are here, something has gone wrong with the database ***/
				$message = $message .'We are unable to process your request. Please try again later.';
				$message = $message.' '.($e->getMessage());
				error_log($file."Error message ".$e->getMessage()." \n", 3, "debug.log");
			}
		}
	
		return json_encode(array(
				'status' => $status,
				'message' => $message
		));
	}
}

?>

File: Eclipse Projects/webapps/TeleDirectJavaServer/src/com/teleDirect/database/DatabaseProvider.java
package com.teleDirect.database;

import java.lang.reflect.Field;
import java.sql.*;
import java.util.Map;

import com.teleDirect.Person;
import com.teleDirect.PersonCollection;
import com.teleDirect.PersonFieldsDb;
import com.teleDirect.PersonFieldsJs;
import com.teleDirect.server.Response;

/**
 * This class is reponsible for all the database operations
 * @author AG041955
 *
 */
public class DatabaseProvider {

	private static final String JDBC_DRIVER = "com.mysql.jdbc.Driver";
	private static final String PARAMS = "?noAccessToProcedureBodies=true";
	private static final String DB_NAME = "directory";
	private static final String DB_URL = "jdbc:mysql://localhost:3306/"+ DB_NAME + PARAMS;

	private static final String USER = "root";
	private static final String PASS = "hello";

	public Connection conn;
	public Statement stmt;

	/**
	 * contructor
	 */
	public DatabaseProvider() {
		conn = null;
		stmt = null;
	}

	/**
	 * @return Returns an active database Connection 
	 */
	public Connection getDatabaseConnection() {
		try {
			// Register JDBC driver
			Class.forName(JDBC_DRIVER);
			
			// Get connection if null
			if(this.conn == null)
				this.conn = DriverManager.getConnection(DB_URL, USER, PASS);
			
		} catch (SQLException e) {
			// Handle errors for JDBC
			e.printStackTrace();
		}catch (ClassNotFoundException e){
			e.printStackTrace();
		}
		return this.conn;
	}


	/**
	 * closes an active database connection
	 * @return void
	 */
	public void closeDatabaseConnection() {
		try {
			if (this.conn != null)
				this.conn.close();
		} catch (SQLException e) {
			e.printStackTrace();
		}
	}
	
	/**
	 * closes an active database connection along with the result set
	 * @return void
	 */
	public void closeDatabaseConnection(ResultSet rs) {
		try {
			if (rs != null)
				rs.close();
			if (this.conn != null)
				this.conn.close();
		} catch (SQLException e) {
			e.printStackTrace();
		}
	}
	
	/**
	 * Used to execute a query statement from the database and get the results
	 * 
	 * @param sql The SQL query string
	 * @return A ResulSet object containing the rows fetched from the database
	 */
	public ResultSet executeQuery(String sql){
		ResultSet rSet = null;
		
		try{
			conn = getDatabaseConnection();
			stmt = conn.createStatement();
			rSet = stmt.executeQuery(sql);
		}catch (SQLException e) {
			e.printStackTrace();
		}catch (Exception e){
			e.printStackTrace();
		}
		
		return rSet;	
	}
	
	/**
	 * Used to execute update statements on the database and make 
	 * appropraite modifications in the database.
	 * 
	 * @param sql SQL update string
	 * @return true/false denoting a successful/unsuccessful execution
	 */
	public boolean executeUpdate(String sql){
		try{
			conn = getDatabaseConnection();
			stmt = conn.createStatement();
			stmt.executeUpdate(sql);
			return true;
		}catch (SQLException e) {
			e.printStackTrace();
		}
		
		return false;
	}
	
	/**
	 * Fetches list of persons from the database and forms a collection of persons
	 * The person collection is added to the data in Response along with an appropriate status and message.
	 * 
	 * @return A Response object contaning the status, message and the PersonCollection
	 */
	public Response fetchPersonList(){
		String status = "", message = ""; 
		Object data = null;
		
		PersonCollection persons = new PersonCollection();
		String sql = generateFetchQuery();
		ResultSet rs = executeQuery(sql);
		int numResults = 0;
		
		try{
			while(rs.next()){
				PersonFieldsJs[] JsFields = PersonFieldsJs.values();
				String[] values = new String[JsFields.length];
				for(int i=0; i<JsFields.length; i++){
					String fieldName = JsFields[i].getFieldName();
					values[i] = rs.getString(fieldName);
					if(fieldName.equals(PersonFieldsJs.DOB.getFieldName()) && values[i] == null)
						values[i] ="";
				}
				Person person = new Person(values);
				persons.add(person);
				++numResults;
			}
			
			if(numResults == 0){
				status = "failure";
				message = "No contacts found";
			}else{
				status = "success";
				message = numResults + " contacts found";
				data = persons;
			}
		}catch (SQLException e){
			status = "error";
			message = e.getMessage() + " We are unable to process your request. Please try again later.";
			e.printStackTrace();
		}
		
		return new Response(status, message, data);
	}
	
	/**
	 * Saves the collection of persons to the database and returns a Response containing an appropriate status and message.
	 * @param persons The colletcion of persons to save to the database
	 * @return A Response object contaning the status and message
	 */
	public Response saveToDatabase(PersonCollection persons){
		String status = "", message = ""; 
		Object data = null;
		
		if(!persons.isEmpty()){
			try{
				Connection conn = getDatabaseConnection();
				
				CallableStatement stmt = conn.prepareCall("{CALL INITIATE_ADD()}");
				stmt.executeQuery();
				
				stmt = conn.prepareCall("{CALL ADD_PERSON(?,?,?,?,?,?,?,?,?,?,?,?,?,?)}");
				
				for (Map.Entry<String, Person> entry : persons.getCollection().entrySet()) {
					Person person = entry.getValue();
					
					int index = 0;
					for (Field f : person.getClass().getDeclaredFields()) {
						if(index!=0){
							if(f.getName().equals(PersonFieldsJs.DOB.getFieldName()) && ((String)f.get(person)).equals("")){
								stmt.setString(index, null);
							}else
								stmt.setString(index, (String)f.get(person));
						}
							
						++index;
					}
					
					ResultSet rs = stmt.executeQuery();
					if(rs != null){
						status = "success";
					}else{
						message += "Could not add : " + stmt.getString(1) + " ";
					}
				}
				message += "Saving Completed";
				
			}catch (SQLException e){
				status = "error";
				message = e.getMessage() + " We are unable to process your request. Please try again later.";
				e.printStackTrace();
			}catch (IllegalAccessException e){
				status = "error";
				message = e.getMessage() + " We are unable to process your request. Please try again later.";
				e.printStackTrace();
			}
		}else{
			status = "error";
			message = "Person Collection is empty";
		}
		return new Response(status, message, data);
	}
	
	/**
	 * Generates a fetch SQL query from the list of fields defined in the
	 * PersonFieldsDb and PersonFieldsJs enums.
	 * 
	 * @return the fetch Sql query
	 */
	public String generateFetchQuery(){
		String selectFields = "";
		PersonFieldsDb[] DbFields = PersonFieldsDb.values();
		PersonFieldsJs[] JsFields = PersonFieldsJs.values();
		
		for(int i=0 ; i<DbFields.length; i++ ){
			if(DbFields[i] == PersonFieldsDb.ID)
				selectFields += "persons." + DbFields[i].getFieldName() + " AS " + JsFields[i].getFieldName() + ", ";
			else
				selectFields += DbFields[i].getFieldName() + " AS " + JsFields[i].getFieldName() + ", ";
		}
		
		// remove last two characters
		selectFields = selectFields.substring(0, selectFields.length()-2);
		
		String sql = 	"SELECT " + selectFields + " FROM persons INNER JOIN personadddetails " + 
						"ON persons." + PersonFieldsDb.ID.getFieldName() + " = personadddetails." + PersonFieldsDb.ID.getFieldName() + 
						" ORDER BY " + PersonFieldsJs.FIRST_NAME.getFieldName();
		return sql;
	}
	
}
